contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,768 @@
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(
+        address indexed _recipient,     // The recipient address of the issued SetTokens
+        ISetToken indexed _setToken,    // The issued SetToken
+        address indexed _inputToken,    // The address of the input asset(ERC20/ETH) used to issue the SetTokens
+        uint256 _amountInputToken,      // The amount of input tokens used for issuance
+        uint256 _amountSetIssued        // The amount of SetTokens received by the recipient
+    );
+
+    event ExchangeRedeem(
+        address indexed _recipient,     // The recipient address which redeemed the SetTokens
+        ISetToken indexed _setToken,    // The redeemed SetToken
+        address indexed _outputToken,   // The addres of output asset(ERC20/ETH) received by the recipient
+        uint256 _amountSetRedeemed,     // The amount of SetTokens redeemed for output tokens
+        uint256 _amountOutputToken      // The amount of output tokens received by the recipient
+    );
+    
+    /* ============ Modifiers ============ */
+    
+    modifier isSetToken(ISetToken _setToken) {
+         require(setController.isSet(address(_setToken)), "ExchangeIssuance: INVALID SET");
+         _;
+    }
+    
+    /* ============ Constructor ============ */
+
+    constructor(
+        address _uniFactory,
+        IUniswapV2Router02 _uniRouter, 
+        address _sushiFactory, 
+        IUniswapV2Router02 _sushiRouter, 
+        IController _setController,
+        IBasicIssuanceModule _basicIssuanceModule
+    )
+        public
+    {
+        uniFactory = _uniFactory;
+        uniRouter = _uniRouter;
+
+        sushiFactory = _sushiFactory;
+        sushiRouter = _sushiRouter;
+
+        WETH = uniRouter.WETH();
+        setController = _setController;
+        basicIssuanceModule = _basicIssuanceModule;
+        IERC20(WETH).safeApprove(address(uniRouter), PreciseUnitMath.maxUint256());
+        IERC20(WETH).safeApprove(address(sushiRouter), PreciseUnitMath.maxUint256());
+    }
+    
+    /* ============ Public Functions ============ */
+    
+    /**
+     * Runs all the necessary approval functions required for a given ERC20 token.
+     * This function can be called when a new token is added to a SetToken during a 
+     * rebalance.
+     *
+     * @param _token    Address of the token which needs approval
+     */
+    function approveToken(IERC20 _token) public {
+        _safeApprove(_token, address(uniRouter));
+        _safeApprove(_token, address(sushiRouter));
+        _safeApprove(_token, address(basicIssuanceModule));
+    }
+
+    /* ============ External Functions ============ */
+    
+    /**
+     * Runs all the necessary approval functions required for a list of ERC20 tokens.
+     *
+     * @param _tokens    Addresses of the tokens which need approval
+     */
+    function approveTokens(IERC20[] calldata _tokens) external {
+        for (uint256 i = 0; i < _tokens.length; i++)
```

Comment:
Knit: Let's try to keep consistency re: for loop braces. The rest of the code-base and this module has braces, so let's be consistent here.